common/djangoapps/student/models.py

```diff
@@ -812,6 +816,16 @@ def user_post_save_callback(sender, **kwargs):
                         cea.save()
                     continue
 
+                # Skip auto enrollment of user if
+                # DISABLE_AUTO_ENROLL_AFTER_ENROLLMENT_END is set to true
+                # and the course enrollment has ended
+                if (
+                    disabled_auto_enroll_after_enrollment_end and
+                    CourseOverview.get_from_id(cea.course_id).enrollment_has_ended()
+                ):
+                    log.info(f'Skipping auto enrollment of user as enrollment for course {cea.course_id} has ended')
+                    continue
```

Comment:
Even if we restrict this here, users are still able to visit the Course About page to enroll themselves.
We could mitigate this by moving these conditions above the `cea.user = user` binding, but it still would not affect users who already have an attached CEA. Example scenario:
1. A non-existent email is enrolled without the "Auto Enroll" option checked.
2. A user registers with this email and confirms it while the enrollment is still open. At this point, the CEA is bound to the user.
3. Enrollment ends, but the user can still visit the Course About page to enroll in the course.

Therefore, I'm suggesting keeping this solution here and also adding this check above [this line](https://github.com/openedx/edx-platform/blob/a4ec4c1b8ebeb3d7920a70d1d6f993fc9fdd1d4f/lms/djangoapps/courseware/access.py#L258). This way, it will affect [student self-enrollment](https://github.com/openedx/edx-platform/blob/b8ecfed67dc0520b8c4d95de3096b35acc083611/common/djangoapps/student/views/management.py#L386).
Alternatively, we could add it above [the auth check](https://github.com/openedx/edx-platform/blob/a4ec4c1b8ebeb3d7920a70d1d6f993fc9fdd1d4f/lms/djangoapps/courseware/access.py#L257). to verify this scenario in the [Enrollment API](https://github.com/openedx/edx-platform/blob/ea3dc9cad6aad0138bed352867a38528e17bf690/openedx/core/djangoapps/enrollments/api.py#L254), but I think this is not something we want to do.

We should also exclude [staff and instructors](https://github.com/openedx/edx-platform/blob/a4ec4c1b8ebeb3d7920a70d1d6f993fc9fdd1d4f/lms/djangoapps/courseware/access.py#L268) from this functionality, as these users should not be restricted to the enrollment period.

Reply:
@Agrendalath 

I missed this scenario, where the invited user could manually enroll even though the user was not auto enrolled. Thanks for pointing this out.

I have accordingly added the check in [assess.py](https://github.com/open-craft/edx-platform/blob/cc700bad8e791ad94df8fb5b93fe2e89b2e2b8ef/lms/djangoapps/courseware/access.py#L259-L274).
However, instead of adding it above [this line](https://github.com/openedx/edx-platform/blob/a4ec4c1b8ebeb3d7920a70d1d6f993fc9fdd1d4f/lms/djangoapps/courseware/access.py#L258), I have refactored the `if cea` condition below and added it inside this condition, since this check is applicable only if there is a `cea` entry against the user's email id.

Also, instead of maintaining the same logic at two places, I have removed the check from [here](https://github.com/open-craft/edx-platform/blob/cc700bad8e791ad94df8fb5b93fe2e89b2e2b8ef/common/djangoapps/student/models.py#L815-L820) and am instead calling [CourseEnrollment.is_enrollment_closed](https://github.com/open-craft/edx-platform/blob/cc700bad8e791ad94df8fb5b93fe2e89b2e2b8ef/common/djangoapps/student/models.py#L1384)

Reply:
> We should also exclude [staff and instructors](https://github.com/openedx/edx-platform/blob/a4ec4c1b8ebeb3d7920a70d1d6f993fc9fdd1d4f/lms/djangoapps/courseware/access.py#L268) from this functionality, as these users should not be restricted to the enrollment period.

@kaustavb12, does it still work for users matching `_has_staff_access_to_descriptor`, though? I haven't tested this yet (I'm reinstalling devstack deps after switching through lots of branches), but [adding them](https://github.com/openedx/edx-platform/blob/068eecff91cbdfc2fe6c6770a7acb3985e7fa654/lms/djangoapps/instructor/views/api.py#L981-L1059) through the `Instructor Dashboard -> Membership -> Course Team Management` doesn't enroll them. If such a user has a CEA (same scenario as above), then they should still be able to enroll themselves in the course. The current code in `access.py` doesn't seem to be covering it.

Reply:
@Agrendalath 
Indeed, this an edge case where the invited user registers after enrollment has ended and is subsequently granted staff access. In this case the user is unable to enroll.

I have now fixed this by adding a check so that the new feature flag does not affect for staff user.